sa/storage-authority.go

```diff
@@ -61,32 +62,70 @@ func (ssa *SQLStorageAuthority) InitTables() (err error) {
 		return
 	}
 
+	// All fields should be created with "NOT NULL" because go's SQL support does not
+	// handle null values well (see, e.g. https://github.com/go-sql-driver/mysql/issues/59)
+	statements := []string{
+
 	// Create registrations table
-	_, err = tx.Exec("CREATE TABLE IF NOT EXISTS registrations (id TEXT, thumbprint TEXT, value TEXT);")
-	if err != nil {
-		tx.Rollback()
-		return
-	}
+	`CREATE TABLE IF NOT EXISTS registrations (
+		id TEXT NOT NULL,
+		thumbprint TEXT NOT NULL,
+		value TEXT NOT NULL
+	);`,
 
 	// Create pending authorizations table
-	_, err = tx.Exec("CREATE TABLE IF NOT EXISTS pending_authz (id TEXT, value BLOB);")
-	if err != nil {
-		tx.Rollback()
-		return
-	}
+	// TODO: Add NOT NULL to value. Right now it causes test failures because some
+	// inserts to not fill all fields.
+	`CREATE TABLE IF NOT EXISTS pending_authz (
+		id TEXT NOT NULL,
+		value BLOB
+	);`,
 
 	// Create finalized authorizations table
-	_, err = tx.Exec("CREATE TABLE IF NOT EXISTS authz (sequence INTEGER, id TEXT, digest TEXT, value BLOB);")
-	if err != nil {
-		tx.Rollback()
-		return
-	}
-
-	// Create certificates table
-	_, err = tx.Exec("CREATE TABLE IF NOT EXISTS certificates (serial TEXT, digest TEXT, value BLOB);")
-	if err != nil {
-		tx.Rollback()
-		return
+	`CREATE TABLE IF NOT EXISTS authz (
+		sequence INTEGER NOT NULL,
+		id TEXT NOT NULL,
+		digest TEXT NOT NULL,
+		value BLOB NOT NULL
+	);`,
+
+	// Create certificates table. This should be effectively append-only, enforced
+	// by DB permissions.
+	`CREATE TABLE IF NOT EXISTS certificates (
+		serial STRING NOT NULL,
```

Comment:
I looked into TEXT vs VARCHAR vs STRING, and it sounds like it's a big complicated, but VARCHAR is probably what we want: https://stackoverflow.com/questions/6628660/text-vs-varchar-in-innodb-mysql-5-5-when-to-use-each-one.

I'm going to use VARCHAR(255) as a default for now since (a) it doesn't affect storage size, (b) it gives us a bit of flexibility in the contents of these fields, and (c) it won't truncate if we go over, so we can find erroneous entries rather than having them truncated into "looks ok."